File: tui/src/views/cluster.zig
const std = @import("std");
const vaxis = @import("vaxis");
const poll = @import("../poll.zig");

const Table = vaxis.widgets.Table;
const Cell = vaxis.Cell;

pub const ClusterView = struct {
    table_ctx: Table.TableContext,
    num_rows: u16 = 0,

    pub fn init() ClusterView {
        return .{
            .table_ctx = .{
                .active = true,
                .selected_bg = .{ .index = 4 },
                .selected_fg = .{ .index = 0 },
                .active_bg = .{ .index = 4 },
                .active_fg = .{ .index = 0 },
                .hdr_bg_1 = .{ .index = 8 },
                .hdr_bg_2 = .{ .index = 8 },
                .row_bg_1 = .default,
                .row_bg_2 = .default,
                .col_width = .dynamic_fill,
                .header_names = .{ .custom = &.{
                    "Name", "Role", "IP", "PVE Node", "VMID", "Talos Ver", "K8s Ver", "Etcd", "Health",
                } },
            },
        };
    }

    pub fn handleKey(self: *ClusterView, key: vaxis.Key) void {
        if (self.num_rows == 0) return;

        if (key.matches('j', .{}) or key.matches(vaxis.Key.down, .{})) {
            if (self.table_ctx.row < self.num_rows - 1) {
                self.table_ctx.row += 1;
            }
        } else if (key.matches('k', .{}) or key.matches(vaxis.Key.up, .{})) {
            if (self.table_ctx.row > 0) {
                self.table_ctx.row -= 1;
            }
        } else if (key.matches('g', .{})) {
            // gg: go to top (single g for now)
            self.table_ctx.row = 0;
        } else if (key.matches('G', .{ .shift = true })) {
            // G: go to bottom
            if (self.num_rows > 0) {
                self.table_ctx.row = self.num_rows - 1;
            }
        }
    }

    pub fn draw(self: *ClusterView, alloc: std.mem.Allocator, win: vaxis.Window, rows: []const poll.NodeRow) void {
        self.num_rows = @intCast(rows.len);
        if (rows.len == 0) {
            self.drawEmpty(win);
            return;
        }

        // Clamp selected row
        if (self.table_ctx.row >= self.num_rows) {
            self.table_ctx.row = self.num_rows - 1;
        }

        Table.drawTable(alloc, win, rows, &self.table_ctx) catch {
            self.drawEmpty(win);
        };
    }

    fn drawEmpty(self: *ClusterView, win: vaxis.Window) void {
        _ = self;
        const msg = "No cluster data available";
        const col: u16 = if (win.width > msg.len) (win.width - @as(u16, @intCast(msg.len))) / 2 else 0;
        const row: u16 = win.height / 2;
        _ = win.print(&.{.{ .text = msg, .style = .{ .fg = .{ .index = 8 } } }}, .{
            .col_offset = col,
            .row_offset = row,
            .wrap = .none,
        });
    }
};

File: tui/src/views/cluster.rs
use crate::models::Snapshot;
use crate::util::truncate;
use crossterm::event::{KeyCode, KeyEvent};
use ratatui::{
    Frame,
    layout::{Constraint, Rect},
    style::{Color, Modifier, Style},
    widgets::{Block, Borders, Cell, Paragraph, Row, Table, TableState},
};

#[derive(Default)]
pub struct ClusterViewState {
    pub table: TableState,
}

impl ClusterViewState {
    pub fn handle_key(&mut self, key: KeyEvent, row_count: usize) {
        if row_count == 0 {
            self.table.select(None);
            return;
        }
        let selected = self.table.selected().unwrap_or(0);
        let next = match key.code {
            KeyCode::Char('j') | KeyCode::Down => selected.saturating_add(1).min(row_count - 1),
            KeyCode::Char('k') | KeyCode::Up => selected.saturating_sub(1),
            KeyCode::Char('g') => 0,
            KeyCode::Char('G') => row_count - 1,
            _ => selected,
        };
        self.table.select(Some(next));
    }
}

pub fn render(frame: &mut Frame, area: Rect, state: &mut ClusterViewState, snapshot: &Snapshot) {
    if snapshot.cluster_rows.is_empty() {
        frame.render_widget(
            Paragraph::new("No cluster data available")
                .block(Block::default().borders(Borders::ALL).title("Cluster")),
            area,
        );
        return;
    }

    let rows = snapshot.cluster_rows.iter().map(|row| {
        Row::new(vec![
            Cell::from(truncate(&row.name, 16)),
            Cell::from(truncate(&row.role, 12)),
            Cell::from(truncate(&row.ip, 16)),
            Cell::from(truncate(&row.pve_node, 12)),
            Cell::from(row.vmid.clone()),
            Cell::from(truncate(&row.talos_version, 12)),
            Cell::from(truncate(&row.kubernetes_version, 12)),
            Cell::from(truncate(&row.etcd, 10)),
            Cell::from(truncate(&row.health, 12)),
        ])
    });

    let table = Table::new(
        rows,
        [
            Constraint::Length(16),
            Constraint::Length(12),
            Constraint::Length(16),
            Constraint::Length(12),
            Constraint::Length(8),
            Constraint::Length(12),
            Constraint::Length(12),
            Constraint::Length(10),
            Constraint::Min(10),
        ],
    )
    .header(
        Row::new(vec![
            "Name", "Role", "IP", "PVE Node", "VMID", "Talos", "K8s", "Etcd", "Health",
        ])
        .style(
            Style::default()
                .fg(Color::Cyan)
                .add_modifier(Modifier::BOLD),
        ),
    )
    .block(Block::default().borders(Borders::ALL).title("Cluster"))
    .row_highlight_style(
        Style::default()
            .bg(Color::Blue)
            .fg(Color::Black)
            .add_modifier(Modifier::BOLD),
    );

    if state.table.selected().is_none() {
        state.table.select(Some(0));
    }
    frame.render_stateful_widget(table, area, &mut state.table);
}
